### PR TITLE
Stmt aggr

### PR DESCRIPTION
**Pull request by gavinyue**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #9834

## What this PR does / why we need it:

### CODE REVIEW

**Comment by matrix-meow** [comment]
@gavinyue Thanks for your contributions!

**Here are review comments for file pkg/util/trace/impl/motrace/Aggr.go:**
The pull request is titled "Stmt aggr" and the body is a template that has not been filled out. The changes made in the pull request are to the file `pkg/util/trace/impl/motrace/Aggr.go`. The changes include adding a new type `Item` and a new type `Aggregator` with methods `AddItem` and `GetResults`. The `Aggregator` type has fields `ctx`, `Grouped`, `WindowSize`, `NewItemFunc`, `UpdateFunc`, and `FilterFunc`. The `AddItem` method adds an item to the `Grouped` map and returns an error if the item is filtered out. The `GetResults` method returns a slice of all the items in the `Grouped` map.

There are a few issues with this pull request that need to be addressed:

1. The title of the pull request is not descriptive enough. It should provide more information about what the changes are and what they aim to achieve.

2. The body of the pull request is a template that has not been filled out. The template should be removed and replaced with a clear and concise explanation of what the changes are and why they are necessary.

3. The `AddItem` method returns `nil, nil` if an item is added successfully. This is not useful information and should be changed to return the added item and any errors that occurred.

4. The `GetResults` method returns a slice of all the items in the `Grouped` map, regardless of whether they have been filtered out or not. This is not desirable behavior and should be changed to only return items that have not been filtered out.

5. The `Grouped` map is not thread-safe, which could lead to race conditions if multiple goroutines access it concurrently. This should be addressed by using a mutex or a concurrent map.

6. The `Item` interface does not have any methods defined on it. This makes it difficult to understand what it is used for and how it should be implemented. Some methods should be defined on the interface to provide more clarity.

7. The `Aggregator` type has fields that are not used in the current implementation (`ctx`, `NewItemFunc`, `UpdateFunc`). These fields should either be removed or used in the implementation.

To optimize the changes made in the pull request, the following suggestions could be considered:

1. Add more tests to ensure that the `Aggregator` type works as expected and that all edge cases are covered.

2. Consider using a concurrent map instead of a regular map to improve performance and avoid race conditions.

3. Consider adding more methods to the `Item` interface to provide more functionality and make it easier to implement.

4. Consider adding more fields to the `Aggregator` type to provide more flexibility and customization options.

**Here are review comments for file pkg/util/trace/impl/motrace/Aggr_test.go:**
The pull request is titled "Stmt aggr" and has an empty body. The changes made in the pull request are in the file `pkg/util/trace/impl/motrace/Aggr_test.go`. The changes include adding a new test function `TestAggregator` that tests the `NewAggregator` function and the `GetResults` function of the `Aggregator` struct. The test function creates some `StatementInfo` instances and adds them to the aggregator. It then gets the results from the aggregator and tests the expected behavior.

There are no major problems with the changes made in the pull request. However, there are a few minor issues that need to be addressed:

1. The title of the pull request is not descriptive enough. It should include more information about what the changes are and why they are being made.

2. The body of the pull request is empty. It should include a description of the changes being made, why they are necessary, and any other relevant information.

3. The test function `TestAggregator` is not well-documented. It should include comments explaining what each part of the test is doing and why it is necessary.

4. The error messages in the test function are not very informative. They should include more details about what went wrong and how to fix it.

To optimize the changes made in the pull request, the following suggestions can be considered:

1. Add more test cases to cover different scenarios and edge cases.

2. Use test fixtures to reduce code duplication and make the tests more modular.

3. Use test assertions to check for specific conditions and behaviors, rather than relying on error messages.

4. Add more comments and documentation to make the code and tests easier to understand and maintain.

**Here are review comments for file pkg/util/trace/impl/motrace/buffer_pipe.go:**
The pull request is titled "Stmt aggr" and the body is incomplete, as it does not specify the type of PR or the issue(s) it fixes. The changes made in the pull request involve adding an aggregator to the `genETLData` function in `buffer_pipe.go`. The aggregator is used to group StatementInfo items by their attributes and aggregate them. The changes include initializing the aggregator, adding StatementInfo items to it, and getting the aggregated results. 

One problem with this pull request is that the body is incomplete and does not provide enough information about the purpose of the changes. The title is also not very descriptive and could be improved to better reflect the changes made. Additionally, the code changes could benefit from more comments to explain the purpose of the code and how it works. 

To address these issues, the pull request author could provide more information in the body of the pull request, including the type of PR and the issue(s) it fixes. The author could also improve the title to better reflect the changes made, such as "Add StatementInfo aggregator to genETLData function in buffer_pipe.go". Finally, the author could add more comments to the code changes to explain the purpose of the code and how it works.

**Here are review comments for file pkg/util/trace/impl/motrace/report_statement.go:**
The pull request is not well described. The title "Stmt aggr" is not informative and does not provide any context about the changes made. The body of the pull request is empty, except for a checklist that has not been filled out. This makes it difficult to understand the purpose of the changes.

The changes made in the pull request are adding new functions and modifying existing ones in the `pkg/util/trace/impl/motrace/report_statement.go` file. The modifications include adding a new struct `Key`, adding a new method `Key` to the `StatementInfo` struct, and modifying the `StatementInfoNew`, `StatementInfoUpdate`, and `StatementInfoFilter` functions.

The modifications seem to be related to aggregating and filtering statement information. However, it is not clear what problem these changes are trying to solve or what benefits they provide. The code changes are also not well documented, making it difficult to understand their purpose and how they work.

To improve the pull request, the title should be more descriptive, and the body should provide a clear explanation of the problem being solved and the benefits of the changes. The modifications should also be better documented, with comments explaining their purpose and how they work. Additionally, the checklist in the body should be filled out to indicate the type of pull request and any issues being fixed.

**Here are review comments for file pkg/util/trace/impl/motrace/report_statement_test.go:**
The pull request title "Stmt aggr" is not informative and does not provide any context about the changes made. It would be helpful to have a more descriptive title that summarizes the purpose of the pull request.

The body of the pull request is incomplete and does not provide any information about the changes made or the reason for making them. It also includes a checklist of PR types, but none of the options are selected. The body should be updated to include a clear description of the changes made and the motivation behind them.

In the changes made to the file `pkg/util/trace/impl/motrace/report_statement_test.go`, there are no issues identified. However, it would be helpful to have a more detailed description of the changes made to the file. The changes appear to be adding imports and reordering them, but it is not clear why these changes were made. It would be helpful to have a comment explaining the purpose of these changes.

Overall, the pull request needs more information to be properly reviewed. The title should be updated to be more descriptive, the body should include a clear description of the changes made and the motivation behind them, and the changes made to the file should be explained in more detail.

**Here are review comments for file pkg/util/trace/impl/motrace/schema.go:**
The pull request title "Stmt aggr" is not descriptive enough and does not provide any context about the changes made. The body of the pull request is incomplete and does not provide any information about the purpose of the changes. It also includes a checklist of PR types, but it is not clear why this is included or how it is relevant to the changes made.

The changes made in the pull request modify the `statsCol` variable in the `schema.go` file. The default value for this variable is changed from an empty object `{}` to an empty array `[]`. However, there is no explanation provided for why this change is necessary or how it will impact the codebase.

Suggestions for improvement:

- Provide a more descriptive and informative title for the pull request that summarizes the changes made.
- Include a detailed description in the body of the pull request that explains the purpose of the changes and how they will impact the codebase.
- Remove the checklist of PR types as it is not relevant to the changes made.
- Consider providing more context about the `statsCol` variable and why the default value is being changed to an empty array. This will help reviewers understand the impact of the changes and ensure that they are necessary.

File: pkg/util/trace/impl/motrace/report_statement.go [hunk 1]
```diff
@@ -36,6 +36,71 @@ var NilTxnID [16]byte
...
+		// The item couldn't be cast to *StatementInfo
+		return false
+	}
+
+	// Check SqlSourceType
+	switch statementInfo.SqlSourceType {
+	case "internal_sql", "external_sql", "non_cloud_user":
+		// Check StatementType
+		switch statementInfo.StatementType {
+		case "Insert", "Update", "Delete", "Execute":
```
Comment (xzxiong) [inline pkg/util/trace/impl/motrace/report_statement.go:90]: inset 、 update、 delete 都直接聚合了？

File: pkg/util/trace/impl/motrace/report_statement.go [hunk 2]
```diff
@@ -36,6 +36,71 @@ var NilTxnID [16]byte
...
+			return true
+		case "Select":
+			// For 'select', also check if Duration is longer than 200 milliseconds
+			if statementInfo.Duration < 200*time.Millisecond {
```
Comment (xzxiong) [inline pkg/util/trace/impl/motrace/report_statement.go:94]: 这里 `200 * time.Millisecond` 应该会改成配置吧？

File: pkg/util/trace/impl/motrace/report_statement.go [hunk 3]
```diff
@@ -36,6 +36,71 @@ var NilTxnID [16]byte
...
+	// Attempt to perform a type assertion to *StatementInfo
+	statementInfo, ok := i.(*StatementInfo)
+
+	if !ok {
+		// The item couldn't be cast to *StatementInfo
+		return false
+	}
+
+	// Check SqlSourceType
+	switch statementInfo.SqlSourceType {
```
Comment (xzxiong) [inline pkg/util/trace/impl/motrace/report_statement.go:86]: 按这里的逻辑，status:Running 会被统计次数 +1

File: pkg/util/trace/impl/motrace/report_statement.go [hunk 4]
```diff
@@ -36,6 +36,71 @@ var NilTxnID [16]byte
...
+		s.StatementFingerprint = ""
+		s.Error = nil
+		s.RowsRead = 0
+		s.BytesScan = 0
+		s.ResultCount = 0
+		return s
+	}
+	return nil
+}
+func StatementInfoUpdate(existing, new Item) {
```
Comment (xzxiong) [inline pkg/util/trace/impl/motrace/report_statement.go:62]: 这里不计数么？不统计聚合的 query个数？

File: pkg/util/trace/impl/motrace/report_statement.go (hunk 1 above)
Comment (xzxiong) [reply pkg/util/trace/impl/motrace/report_statement.go:91]: https://doc.weixin.qq.com/doc/w3_AZUAugYIAOcKdTPxcOmSoSBADCVSM?scode=AJsA6gc3AA8dS9cv05AW0A-gb6AOI

File: pkg/util/trace/impl/motrace/report_statement.go [hunk 5]
```diff
@@ -36,6 +37,71 @@ var NilTxnID [16]byte
...
+
+	e := existing.(*StatementInfo)
+	n := new.(*StatementInfo)
+	// update the stats
+	e.Duration += n.Duration
+	e.Statement += n.Statement + "; "
```
Comment (fengttt) [inline pkg/util/trace/impl/motrace/report_statement.go:69]: Don't do this, just update a counter.    Each statement is not interesting.

File: pkg/util/trace/impl/motrace/report_statement.go [hunk 6]
```diff
@@ -36,6 +37,71 @@ var NilTxnID [16]byte
...
+	// reponseAt is the last response time
+	e.ResponseAt = n.ResponseAt
+	// TODO: update the stats still json here
+	n.ExecPlan2Json(context.Background())
+	// e.statsJsonByte
```
Comment (fengttt) [inline pkg/util/trace/impl/motrace/report_statement.go:74]: Why do we still need ExecPlan2Json?   We must remove this.      All we need is update statement stats, which should direct read from memory, no to/from json.

**Review by fengttt — CHANGES_REQUESTED** [review]
CIL.

File: pkg/util/trace/impl/motrace/report_statement.go (hunk 6 above)
Comment (fengttt) [reply pkg/util/trace/impl/motrace/report_statement.go:74]: Way too expensive.    All you need is to get 2 or 3 numbers out of this thing.

**Review by fengttt — CHANGES_REQUESTED** [review]
schema.go,

I don't think we should change stats to an array, just aggregate the stats if you can. 

Should add one column as statement count.

**Comment by gavinyue** [comment]
> schema.go,
> 
> I don't think we should change stats to an array, just aggregate the stats if you can.
> 
> Should add one column as statement count.

Add one column for the count, no problem. 


Stats right now is a text field, which contains a complex JSON.  I follow this doc to change it the array. https://doc.weixin.qq.com/doc/w3_AW0A-gb6AOIPqDxGO0tTMeyjnIZrk?scode=AJsA6gc3AA8pSn6FRZAS0AvwapAOk

Otherwise it would be a simplified JSON?

**Comment by fengttt** [comment]
> > schema.go,
> > I don't think we should change stats to an array, just aggregate the stats if you can.
> > Should add one column as statement count.
> 
> Add one column for the count, no problem.
> 
> Stats right now is a text field, which contains a complex JSON. I follow this doc to change it the array. https://doc.weixin.qq.com/doc/w3_AW0A-gb6AOIPqDxGO0tTMeyjnIZrk?scode=AJsA6gc3AA8pSn6FRZAS0AvwapAOk
> 
> Otherwise it would be a simplified JSON?

Ah that is ok -- i was mistaken, thinking you will not aggregate stats but use an array of json struct.     Thank you

File: pkg/util/trace/impl/motrace/report_statement.go (hunk 4 above)
Comment (gavinyue) [reply pkg/util/trace/impl/motrace/report_statement.go:67]: 多价格一个column 来统计聚合的数量了

File: pkg/util/trace/impl/motrace/report_statement.go (hunk 3 above)
Comment (gavinyue) [reply pkg/util/trace/impl/motrace/report_statement.go:103]: filter out the running

File: pkg/util/trace/impl/motrace/report_statement.go (hunk 2 above)
Comment (gavinyue) [reply pkg/util/trace/impl/motrace/report_statement.go]: Just changed

File: pkg/util/trace/impl/motrace/report_statement.go [hunk 7]
```diff
@@ -207,9 +301,105 @@ func (s *StatementInfo) FillRow(ctx context.Context, row *table.Row) {
 	row.SetColumnVal(bytesScanCol, table.Int64Field(s.BytesScan))
 	row.SetColumnVal(stmtTypeCol, table.StringField(s.StatementType))
 	row.SetColumnVal(queryTypeCol, table.StringField(s.QueryType))
+	row.SetColumnVal(aggrCntCol, table.Int64Field(s.AggrCount))
 	row.SetColumnVal(resultCntCol, table.Int64Field(s.ResultCount))
 }
 
+func getStatsValues(jsonData []byte, duration uint64) ([]uint64, error) {
```
Comment (xzxiong) [inline pkg/util/trace/impl/motrace/report_statement.go:308]: Do Unmarshal cost too high

just get the array from mysql_cmd_executor.go, func marshalPlanHandler.Marshal

you can change the function  signature as you need.
- from `Marshal(ctx context.Context) (jsonBytes []byte, statsJonsBytes []byte, stats motrace.Statistic)`
- to `Marshal(ctx context.Context) (jsonBytes []byte, stats []int, stats motrace.Statistic)`

甚至于应该实现一个 struct [`StatsArray`](https://doc.weixin.qq.com/doc/w3_AW0A-gb6AOIi8I05LwBRWe32S3h18?scode=AJsA6gc3AA8CC2VgFkAW0A-gb6AOI), which is defined in [doc](https://doc.weixin.qq.com/doc/w3_AW0A-gb6AOIPqDxGO0tTMeyjnIZrk?scode=AJsA6gc3AA8VyTaLZKAW0A-gb6AOI)

**Review by fengttt — APPROVED** [review]
Please address the review points.

File: pkg/util/trace/impl/motrace/report_statement.go (hunk 6 above)
Comment (gavinyue) [reply pkg/util/trace/impl/motrace/report_statement.go]: Resolved. Do not need the json anymore

File: pkg/util/trace/impl/motrace/report_statement.go (hunk 7 above)
Comment (gavinyue) [reply pkg/util/trace/impl/motrace/report_statement.go]: Resolved. Do not need the json anymore

**Review by aressu1985 — APPROVED** [review]

**Review by daviszhen — APPROVED** [review]

File: test/distributed/cases/statement_query_type/statement_query_type.result [hunk 8]
```diff
@@ -257,26 +253,19 @@ show create database test_db    Other    external_sql
 create view test_view as select * from test_table    DDL    external_sql
 create table test_table(\ncol1 int,\ncol2 varchar\n)    DDL    external_sql
 use test_db    Other    external_sql
-create database test_db    DDL    external_sql
 begin    TCL    external_sql
+create database test_db    DDL    external_sql
```
Comment (xzxiong) [inline test/distributed/cases/statement_query_type/statement_query_type.result:257]: i confuse. why this stmt changes the order ?

Comment (gavinyue) [reply test/distributed/cases/statement_query_type/statement_query_type.result:257]: Becasue the aggr,  some stmts, the one that should be aggreagted, will be put into a window first.  And will be written down at that end. 

So the order will be modified.

File: test/distributed/cases/log/query_version.result [hunk 9]
```diff
@@ -1,14 +1,17 @@
...
 use mysql; select * from user limit 0;
 select sleep(15) as s;
 s
 0
 select count(1) as cnt, statement_id, statement, status from system.statement_info group by statement_id, statement, status having count(1) > 1;
 cnt    statement_id    statement    status
-select `database` from system.statement_info where statement = 'select * from user limit 0' and request_at > @ts order by request_at desc limit 1;
-database
-mysql
+select `database`, `aggr_count` from system.statement_info where statement LIKE '%select * from user limit 0%' order by request_at desc limit 1;
```
[the remark this answers is not in the document]
Comment (gavinyue) [reply test/distributed/cases/log/query_version.result:15]: because the window,   the statement we are looking for is aggregated with the one before the @ts and uses that one's request_at timestamp.

File: test/distributed/cases/result_count/result_count.sql [hunk 10]
```diff
@@ -117,7 +117,7 @@ use system;
 -- For instance, 'set @a = (select 1)'. The ast of '(select 1)' will be rewritten to the ast of 'select (select 1) from dual'.
 -- We do not have the source sql of 'select (select 1) from dual'. We just have the ast of 'select (select 1) from dual'.
 -- So the statement_info.statement for 'select (select 1) from dual' only is empty string.
-select statement, result_count from statement_info where account="bvt_result_count" and statement not like '%mo_ctl%' and length(trim(statement)) > 0 and status != 'Running' order by request_at desc limit 70;
+select statement, result_count from statement_info where account="bvt_result_count" and statement not like '%mo_ctl%' and length(statement) > 0 and status != 'Running' and aggr_count < 1 order by request_at desc limit 70;
```
[the remark this answers is not in the document]
Comment (gavinyue) [reply test/distributed/cases/result_count/result_count.sql:120]: But in this test,  we have two "select from satement_info", one is for cloud_user and this one is for not -cloud_user. 

Changing every SQL to cloud_user would make these two tests identical.

File: test/distributed/cases/statement_query_type/statement_query_type.result (hunk 8 above)
Comment (xzxiong) [reply test/distributed/cases/statement_query_type/statement_query_type.result:257]: i aware this rule.
In here, this two sql is  `begin` and `create`, both are go into  window.  is that right ?

File: pkg/util/trace/impl/motrace/report_statement.go (hunk 5 above)
Comment (xzxiong) [reply pkg/util/trace/impl/motrace/report_statement.go]: @gavinyue Has this one been solved ? this one may will affect the aggr improvement.

File: test/distributed/cases/result_count/result_count.sql (hunk 10 above)
[the remark this answers is not in the document]
Comment (xzxiong) [reply test/distributed/cases/result_count/result_count.sql:120]: i see. `cloud_user` part already covert those `select/delete/...` case.

File: test/distributed/cases/statement_query_type/statement_query_type.result (hunk 8 above)
Comment (xzxiong) [reply test/distributed/cases/statement_query_type/statement_query_type.result:257]: my fail. current result is the corect one.

**Review by xzxiong — APPROVED** [review]
lgtm, as defined in [【企微文档】【PRD】 查询历史合并](
https://doc.weixin.qq.com/doc/w3_AZUAugYIAOcKdTPxcOmSoSBADCVSM?scode=AJsA6gc3AA8LVUkmrJAZUAugYIAOc)

the further issue should discuss between agg stmt record and Running Record

**Review by qingxinhome — APPROVED** [review]

**Review by heni02 — APPROVED** [review]
